### PR TITLE
Add LGPE+ Premier Ball Bonus config

### PR DESCRIPTION
**Pull request by AsparagusEduardo**
## Description
Originally based off [pret's tutorial](https://github.com/pret/pokeemerald/wiki/LGPE-Style-Bonus-Premier-Balls) but now with extra features:
- If there's no space for the full amount of Premier Balls that would've been given, give the partial amount.
- Shows the amount of Premier Balls given.

## Images
mGBA_6Cm60tBtRg

## **Discord contact info**
AsparagusEduardo

### CODE REVIEW

File: src/shop.c [hunk 1]
```diff
@@ -1162,7 +1162,9 @@ static void Task_ReturnToItemListAfterItemPurchase(u8 taskId)
         PlaySE(SE_SELECT);
 
         // Purchasing 10+ Poke Balls gets the player a Premier Ball
-        if (tItemId == ITEM_POKE_BALL && tItemCount >= 10 && AddBagItem(ITEM_PREMIER_BALL, 1) == TRUE)
+        if (I_PREMIER_BALL_BONUS >= GEN_7 && (ItemId_GetPocket(tItemId) == POCKET_POKE_BALLS) && tItemCount >= 10 && AddBagItem(ITEM_PREMIER_BALL, tItemCount / 10) == TRUE)
+            BuyMenuDisplayMessage(taskId, (tItemCount >= 20 ? gText_ThrowInPremierBalls : gText_ThrowInPremierBall), BuyMenuReturnToItemList);
```
Comment (Bassoonian) [inline src/shop.c:1166]: Rather than just aborting giving 20 Premier Balls if you can only carry 19 more, shouldn't we just give the highest amount possible?

Comment (AsparagusEduardo) [reply src/shop.c:1166]: I think I need to create a method for getting total amount of free space for an item.

Comment (AsparagusEduardo) [reply src/shop.c:1166]: Actually, do we even know what happens in modern games? According to the vanilla Gen 3 code, it won't give you the Premier Ball if there's no space.

Let me boot Scarlet.

Comment (AsparagusEduardo) [reply src/shop.c:1166]: Ok, here are some notes:
- The amount of Premier Balls received is shown in the message.
- If your purchase surpasses the amount of Premier Balls, you get maxed out.
- The message shows up how many Premier Balls you *would've gotten*, not how many you gained.
- The message still shows up if you buy at 999 Premier Balls.

**Comment by AsparagusEduardo** [comment]
Ready for re-review.

File: src/shop.c [hunk 2]
```diff
@@ -1159,11 +1159,26 @@ static void Task_ReturnToItemListAfterItemPurchase(u8 taskId)
 
     if (JOY_NEW(A_BUTTON | B_BUTTON))
     {
-        PlaySE(SE_SELECT);
```
Comment (Bassoonian) [inline src/shop.c:1162]: Why was this removed?

Comment (AsparagusEduardo) [reply src/shop.c:1162]: By error ^^;

File: src/shop.c [hunk 3]
```diff
@@ -1159,11 +1159,26 @@ static void Task_ReturnToItemListAfterItemPurchase(u8 taskId)
...
+        u16 premierBallsToAdd = tItemCount / 10;
+        if (premierBallsToAdd >= 1
+            && ((I_PREMIER_BALL_BONUS <= GEN_6 && tItemId == ITEM_POKE_BALL)
+             || (I_PREMIER_BALL_BONUS >= GEN_7 && (ItemId_GetPocket(tItemId) == POCKET_POKE_BALLS))))
```
[the remark this answers is not in the document]
Comment (AsparagusEduardo) [reply src/shop.c:1165]: If you mean moves, the main reason is that their PLA effects mostly apply to only that game's mechanics. Everything else, afaik, a feature introduced in LGPE is taken as GEN_7+

[the remark this answers is not in the document]
Comment (Bassoonian) [reply src/shop.c:1165]: I'm fairly sure there's exceptions to that (and I would personally disagree with that opinion since for me Gen7+ is primarily Alola and the main-games-but-not-really LGPE), but that's a conversation to be had in Discord at another time

**Comment by AsparagusEduardo** [comment]
Finally addressed comments.

File: src/shop.c [hunk 4]
```diff
@@ -1159,13 +1159,31 @@ static void Task_ReturnToItemListAfterItemPurchase(u8 taskId)
 
     if (JOY_NEW(A_BUTTON | B_BUTTON))
     {
-        PlaySE(SE_SELECT);
+        u16 premierBallsToAdd = tItemCount / 10;
+        if (premierBallsToAdd >= 1
+            && ((I_PREMIER_BALL_BONUS <= GEN_7 && tItemId == ITEM_POKE_BALL)
+             || (I_PREMIER_BALL_BONUS >= GEN_8 && (ItemId_GetPocket(tItemId) == POCKET_POKE_BALLS))))
```
Comment (Bassoonian) [inline src/shop.c:1165]:
```suggestion
        if (premierBallsToAdd >= 1
         && ((I_PREMIER_BALL_BONUS <= GEN_7 && tItemId == ITEM_POKE_BALL)
          || (I_PREMIER_BALL_BONUS >= GEN_8 && (ItemId_GetPocket(tItemId) == POCKET_POKE_BALLS))))
```

**Review by Bassoonian — CHANGES_REQUESTED** [review]

**Review by Bassoonian — APPROVED** [review]
